influent-transport/src/main/java/influent/internal/nio/NioSslAcceptor.java

```diff
@@ -0,0 +1,215 @@
+      serverSocketChannel.setOption(StandardSocketOptions.SO_REUSEADDR, true);
+      serverSocketChannel.bind(this.localAddress, backlog);
+      if (receiveBufferSize > 0) {
+        serverSocketChannel
+            .setOption(StandardSocketOptions.SO_RCVBUF, receiveBufferSize);
+      }
+      eventLoop.register(serverSocketChannel, SelectionKey.OP_ACCEPT, this);
+      logger.info("A NioSslAcceptor is bound with {}.", localAddress);
+    } catch (final AlreadyBoundException | UnsupportedAddressTypeException e) {
+      throw new IllegalArgumentException(e);
+    } catch (final ClosedChannelException
+        | SecurityException
+        | UnsupportedOperationException
+        | IllegalArgumentException e) {
+      throw new AssertionError(e);
+    } catch (final IOException e) {
+      final String message = "An IO error occurred. local address = " + this.localAddress;
+      throw new InfluentIOException(message, e);
+    }
+  }
+
+  @Override
+  public void onAcceptable(final SelectionKey key) {
+    logger.debug("New connection request.");
+    while (true) {
+      SocketChannel channel = accept(key);
+      if (channel == null) {
+        break;
+      }
+      callback.accept(channel);
+    }
+  }
+
+  @Override
+  public void close() {
+    Exceptions.ignore(serverSocketChannel::close,
+        "The acceptor bound with " + localAddress + " closed.");
+    logger.info("The acceptor bound with {} closed.", localAddress);
+  }
+
+  private SocketChannel accept(SelectionKey key) {
+    logger.debug("New connection request.");
+    logger.info("New connection request.");
+    try {
+      SocketChannel channel = ((ServerSocketChannel) key.channel()).accept();
+      channel.configureBlocking(false);
+
+      SSLEngine engine = context.createSSLEngine();
+      engine.setUseClientMode(false);
+      engine.beginHandshake();
+
+      if (doHandshake(channel, engine)) {
+        return channel;
+      } else {
+        channel.close();
+        logger.debug("Connection closed due to handshake failure.");
+        logger.info("Connection closed due to handshake failure.");
+      }
+    } catch (IOException e) {
+      logger.error(e.getLocalizedMessage());
+      e.printStackTrace();
+      throw new InfluentIOException(e.getMessage(), e);
+    }
+    return null;
+  }
+
+  private boolean doHandshake(SocketChannel socketChannel, SSLEngine engine) throws IOException {
+    logger.debug("Handshaking...");
+    logger.info("Handshaking...");
+    SSLEngineResult result;
+    SSLEngineResult.HandshakeStatus status;
+    int bufferSize = engine.getSession().getApplicationBufferSize();
+    ByteBuffer buffer = ByteBuffer.allocate(bufferSize);
+    ByteBuffer unwrappedBuffer = ByteBuffer.allocate(bufferSize);
+    ByteBuffer wrappedBuffer = ByteBuffer.allocate(bufferSize);
+
+    status = engine.getHandshakeStatus();
+    logger.info("{}", status);
+    while (status != SSLEngineResult.HandshakeStatus.FINISHED &&
```

Comment:
## I/O model

Please assume that we are using non-blocking I/O.
When non-blocking I/O, `SocketChannel#read` and `SocketChannel#write` do nothing and immediately returns 0 if I/O operations are not available.
For example, following situations.
- A server had consume all the packet from a client and the client has not sent new packets
- A client is too slow and so server's socket buffer is full

Influent is also using `Selector` that informs us of I/O available socket(I/O multiplexing).
https://github.com/okumin/influent/blob/master/influent-transport/src/main/java/influent/internal/nio/NioEventLoopTask.java#L123

These combination enables few thread to handle many connections and to use CPU efficiently.

## SSLEngine

`SSLEngine` has two roles.

- handshake
- encryption/decryption

SSL/TLS depends TCP/IP, so `SSLEngine` depends on `SocketChannel`.
At handshake phase, `SSLEngine#wrap` generates handshake data and `SSLEngine#unwrap` consumes handshake data.
These handshake data are sent/received by non-blocking `SocketChannel#write/read`.
So the following `serverChannel.read(buffer)` does not consume and continues to spin until the peer makes progress and sends data.
To speak of extremes, when the peer hangs, this loop spins forever.

```java
while (isCompleted) {
  switch (engine.getHandshakeStatus()) {
    case NEED_UNWRAP:
      socketChannel.read(buffer);
      // use buffer if data is received
  }
}
```

## Solution

If we would like to some I/O operations, we should register the channel and operations to `Selector` via NioEventLoop.
https://github.com/okumin/influent/blob/master/influent-transport/src/main/java/influent/internal/nio/NioEventLoop.java#L117-L137

For instance, `NioForwardConnection` registers write I/O operation when it want to send ack response.
https://github.com/okumin/influent/blob/master/influent-java/src/main/java/influent/forward/NioForwardConnection.java#L178

Once registering a channel and operations to `NioEventLoop`, `NioEventLoop` will invoke a proper callback method when the operation becomes available.
https://github.com/okumin/influent/blob/master/influent-transport/src/main/java/influent/internal/nio/NioAttachment.java#L35-L73

I think that `Acceptor` should handle TCP level acceptances and leave SSL/TLS handshake to individual connections.
But this may require many changes……